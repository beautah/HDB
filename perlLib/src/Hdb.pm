# package to generically interface with HDB

=head1 NAME

HDB Generic PERL interface

=head1 VERSION

Documentation for Hdb.pm $ID: $

=head1 SYNOPSIS

To create a DBI handle to HDB as app_user, must have the proper environmental
variables set:

 use Hdb;

 my $hdb = Hdb->new;

 $hdb->connect_to_db($dbname, $user, $pass);

To allow editting of values:

 $hdb->set_approle();

Look up ids from database:

 $hdb->get_app_ids(\%nameid);

 Argument is a hash of hashes, with values for the key 'name' for the keys:
 agen, collect, load_app, method, computation;

To get at the DBI handle, 
 my $dbh = Hdb->dbh;

=head1 DESCRIPTION

The Hdb module encapsulates some of the common functionality needed by perl
progams which interact with the database.

=cut

package Hdb;
use Exporter ();
@ISA = qw(Exporter);
@EXPORT = qw(hdbdie connect_to_db set_approle get_app_ids get_SDI);

use DBI;

#$passw_dbh, $dbh, %objslotsdi, $dbname;

sub new {
  my $proto = shift;
  my $class = ref($proto) || $proto;
  my $self  = {};
  $self->{dbh} = undef;
  $self->{dbname} = undef;
  bless ($self, $class);
  return $self;
}

sub DESTROY
{
  my $self = shift;
  if (defined($self->{dbh})) {
    $self->{dbh}->rollback;
    $self->{dbh}->disconnect;
  }
  undef $self;
}

sub dbh
{
  my $self = shift;
  return $self->{dbh};
}

sub hdbdie {
  my $self = shift;
  if (defined($self->{dbh}) and $self->{dbh}->ping) {
    $self->{dbh}->rollback;
    $self->{dbh}->disconnect;
  }
  die @_;
}

# HDB does something interesting with the app_user account
# app_user has the permission to insert into the database, but only
# if the app_role is enabled, which is not by default
# the applications get around this by connecting to the database as 
# psswd_user (from the environment) and getting the password to 
# turn on this role, and then enabling this role.

sub set_approle {
  my $self = shift;

  unless (defined($self->{dbh}) and $self->{dbh}->ping) {
    print "ERROR: attempted to set app role on unconnected DB handle.\n";
    die @_;
  }

  my $passd_dbh;
  my ($approlepass, $approle);
  my ($psswduser, $psswd_statement, $psswdsth);

  sub passdie {
    $psswd_dbh->disconnect;
    die @_;
  }

  if (!defined($psswduser = $ENV{PSSWD_USER})) {
    die "Environment variable PSSWD_USER not set...\n";
  }
  if (!defined($approle = $ENV{APP_ROLE})) {
    die "Environment variable APP_ROLE not set...\n";
  }

  # create connections, exceptions on error, no autocommit

  $psswd_dbh = DBI->connect("dbi:Oracle:$self->{dbname}", $psswduser,
			    $self->{dbname},
			    {RaiseError => 1, AutoCommit => 0 })
                           or passdie $DBI::errstr;

  $psswd_statement =
     "select psswd from role_psswd where UPPER(ROLE) = UPPER('$approle')";

  $psswdsth = $psswd_dbh->prepare($psswd_statement)
                          or passdie $psswdsth->errstr;
  $psswdsth->execute or passdie $psswdsth->errstr;

  $psswdsth->bind_col(1,\$approlepass);
  $psswdsth->fetch;
  $psswdsth->finish;
  $psswd_dbh->disconnect;

  $self->{dbh}->do("set role $approle identified by $approlepass")
                 or hdbdie $DBI::errstr;

  return $self->{dbh};
}


#this function connects to the database as the specified user
sub connect_to_db {
  my $self = shift;
  $self->{dbname} = $_[0];
  my $user = $_[1];
  my $pass=$_[2];

  $self->{dbh} = DBI->connect("dbi:Oracle:$self->{dbname}", $user, $pass,
			     {RaiseError => 1, AutoCommit => 0 })
                     or hdbdie $DBI::errstr;

  return $self->{dbh};
}

sub get_app_ids
{
# Get ids to describe where data came from
  my $self = shift;

  my $nameid = $_[0];

  my $max_len = 11;

  my $sth;

  $sth = $self->dbh->prepare(q{
     BEGIN
           lookup_application(?,?,?,?,?,  /* agen, collect, load_app, method, computation names*/
                              ?,?,?,?,?); /* agen, collect, load_app, method, computation ids */
     END;
   });

  $sth->bind_param(1,$nameid->{agen}->{name});
  $sth->bind_param(2,$nameid->{collect}->{name});
  $sth->bind_param(3,$nameid->{load_app}->{name});
  $sth->bind_param(4,$nameid->{method}->{name});
  $sth->bind_param(5,$nameid->{computation}->{name});
  $sth->bind_param_inout(6, \$nameid->{agen}->{id}, $max_len);
  $sth->bind_param_inout(7, \$nameid->{collect}->{id}, $max_len);
  $sth->bind_param_inout(8, \$nameid->{load_app}->{id}, $max_len);
  $sth->bind_param_inout(9, \$nameid->{method}->{id}, $max_len);
  $sth->bind_param_inout(10, \$nameid->{computation}->{id}, $max_len);

  eval {
    $sth->execute;
  };

  if ($@ or
      !defined($nameid->{agen}->{id}) or
      !defined($nameid->{collect}->{id}) or
      !defined($nameid->{load_app}->{id}) or
      !defined($nameid->{method}->{id}) or
      !defined($nameid->{computation}->{id})) { # something screwed up
    print $self->dbh->errstr, " $@\n";
    hdbdie "Errors occurred during selection of ids for application.\n";
  }

  return $nameid;
}

sub get_SDI
{
  my ($cur_site, $datatype) = $_[0,1];
  my ($sth, $site_datatype);

  my $get_site_datatype_statement = "select site_datatype_id from hdb_site_datatype where site_id = ? and datatype_id = ? ";

  $sth = $hdb->dbh->prepare($get_site_datatype_statement) || die $sth->errstr;
  $sth->bind_param(1,$cur_site);
  $sth->bind_param(2,$datatype);
  $sth->execute || hdbdie $sth->errstr;
  $sth->bind_col(1,\$site_datatype);
#should check here if there are any rows. if not, no SDI exists for this combo.
  $sth->fetch;

  if (!defined($site_datatype)) {
    hdbdie "site datatype id not found for site: $cur_site datatype: $datatype");
  }

  $sth->finish;

  return $site_datatype;

}

1;

