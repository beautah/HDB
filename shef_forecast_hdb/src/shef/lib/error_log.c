   
#include  <stdio.h>
#include  <stdlib.h>
#include  <time.h>
#include  <string.h>


int error_log(char *pname,FILE *fp_log,int pflag,int eflag,
          char *s1,char *s2,char *s3,char *s4,char *s5)


{


	unsigned char mon_name[12][4] = {
		{"Jan"}, {"Feb"}, {"Mar"}, {"Apr"}, {"May"}, {"Jun"},
		{"Jul"}, {"Aug"}, {"Sep"}, {"Oct"}, {"Nov"}, {"Dec"}
	};

    char buf[6][80],emsg[80];
	time_t  now;
	struct tm *tp;
    signed long offset;
    int i;


/*
    if ( fp_log == NULL )
         return(1);
*/
	/* Get time of error message */

	now = time(NULL);
	tp  = gmtime(&now);

    for ( i=0; i < 6; i++)
          buf[i][0]=0;

    if(s1 != NULL)
              strcpy(buf[0],s1);  

                  
    if(s2 != NULL)

                 strcpy(buf[1],s2);
                 

    if(s3 != NULL) 

                 strcpy(buf[2],s3);

                

    if(s4 != NULL) 
 
                 strcpy(buf[3],s4);

                
    if(s5 != NULL) 

                 strcpy(buf[4],s5);

                

	/* Format error message */

	sprintf(emsg, "\n\n%3s %2d %2d %2d:%02d  program:%-16s",
		mon_name[tp->tm_mon], tp->tm_mday, tp->tm_year,
		tp->tm_hour, tp->tm_min, pname);


    /*  write to standard error device if true */
    if ( pflag == 1 )
    {    
       fprintf(stdout,"\n%s",emsg);

       i = 0;
       while ( buf[i][0] != 0 )
       {
          fprintf(stdout,"\n    %s",buf[i]);
          i++;
       }
    }

   /*  write to log if you have a file pointer */
    if ( fp_log != NULL )
    {
       fprintf(fp_log,"\n%s",emsg);

       i = 0;
       while ( buf[i][0] != 0 )
       {
          fprintf(fp_log,"\n    %s",buf[i]);
          i++;
       }
   
    }


    if ( eflag )
    {
       if ( fp_log )
          fclose(fp_log);
       exit(1);
    }
  
    return (1);

}                                                     
