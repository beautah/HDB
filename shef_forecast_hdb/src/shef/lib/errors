Undefined			first referenced
 symbol  			    in file
datet_                              blockdata.o
dotee_                              shdece.o
stats_                              shdrive.o
econtinuation                       dotepre.o
codes_                              blockdata.o
dotend                              dotepre.o
tempfiles_                          dotepre.o
sendflg_                            shbfr.o
sqlerrors                           shefclose.o
test_                               shdeca.o
DOTEPRE                             nextch.o
days_                               blockdata.o
parms2_                             checkparm.o
fp_postlog                          shefclose.o
luns_                               shdecb.o
pname                               stats_utilities.o
parms1_                             checkparm.o
dotbee_                             shdecb.o
datrel_                             shdeca.o
files_                              shef_test.o
cont_                               shdeca.o
xchar_                              nextch.o
err1_                               blockdata.o
err2_                               blockdata.o
testfile                            shefclose.o
comment_                            nextch.o
efound                              dotepre.o
err3_                               blockdata.o
err4_                               blockdata.o
err5_                               blockdata.o
format_                             shdeca.o
err6_                               blockdata.o
fp_                                 getparm.o
err7_                               blockdata.o
parms3_                             checkparm.o
buffer_                             dotepre.o
dotbegin                            dotepre.o
err8_                               blockdata.o
bbuf1_                              shbfr.o
bbuf2_                              shbfr.o
error_                              shbfw.o
bbuf3_                              shbfr.o
main                                /usr/local/lib/gcc-lib/sparc-sun-solaris2.7/2.8.1/crt1.o
sensors                             shefclose.o
PARTIAL_ERROR                       nextch.o
datim_                              shdeca.o
DEBUG                               stats_utilities.o
dtype_                              shdece.o
data_                               shdeca.o
durprob_                            shdeca.o
shefpass                            shdeca.o
ld: fatal: Symbol referencing errors. No output written to test.exe
