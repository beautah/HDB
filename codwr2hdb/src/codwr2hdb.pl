#!/usr/bin/perl -w

#insert HDB library

use lib "$ENV{HDB_ENV}/perlLib/lib";

# don't need this?
use lib "$ENV{HDB_ENV}/perlLib/lib/sun4-solaris";

use LWP::UserAgent;
use Date::Calc qw(Delta_DHMS Decode_Date_EU Today Add_Delta_Days);
use File::Basename;
use Data::Dumper;
use HTML::TableExtract;
use HTML::TokeParser;

use Hdb;

use strict;

use English qw( -no_match_vars );

#check to see command line usage.
my $progname = basename($0);
chomp $progname;

#the following are global variables, these are set by command line arguments
my $insertflag = 1;
my ( $readfile, $runagg, $printurl, $debug, $accountfile );
my ( $hdbuser, $hdbpass, @site_num_list );

#HDB database interface
# global because used in several sub routines
my $hdb;

#global datasource name, used in get_app_ids
my $datasource = "Colorado Division of Water Resources (CODWR)";

#global variables read from database in get_app_ids
my ( $load_app_id, $agen_id, $validation, $url, $collect_id );
my $agen_abbrev = "CODWR";

#global hash reference containing meta data about sites
# initialized by get_codwr_sites below, and referenced throughout
my ($codwr_sites);

#global date string representations for use in retrieving data
my ( $begindatestr, $enddatestr ) = process_args(@ARGV);

$hdb = Hdb->new();
if ( defined($accountfile) ) {
  $hdb->connect_from_file($accountfile);
} else {
  my $dbname;
  if ( !defined( $dbname = $ENV{HDB_LOCAL} ) ) {
    $hdb->hdbdie("Environment variable HDB_LOCAL not set...\n");
  }

  #create connection to database
  $hdb->connect_to_db( $dbname, $hdbuser, $hdbpass );
}

#set write ability on connection (app_role)
$hdb->set_approle();

#Set date format to the format used by their website
$hdb->dbh->do("ALTER SESSION SET NLS_DATE_FORMAT = 'YYYY-MM-DD HH24:MI'")
  or die "Unable to set NLS_DATE_FORMAT to ISO \n";

get_app_ids();

if (@site_num_list) {
  $codwr_sites = get_codwr_sites(@site_num_list);
} else {
  $codwr_sites = get_codwr_sites();
  (@site_num_list) = build_site_num_list($codwr_sites);
}

=head1 FORMAT

The data is expected in tab delimited format. Example:

#---------------------------- Provisional Data -------------------------------
#This system is maintained by the Colorado Division of Water Resources.
#Contact: Colorado Division of Water Resources (303) 866-3581
#
#All data presented on the Colorado Surface Water Conditions web site are 
#provisional and subject to revision. Data users are cautioned to consider
#carefully the provisional nature of the information before using it for 
#decisions that concern personal or public safety or the conduct of business
#that involves substantial monetary or operational consequences.
#
#Data is returned in TAB delimited format. Data miners may find help on automating
#queries and formatting parameters at http://www.dwr.state.co.us/help
##Gaging Station: FLORIDA RIVER BELOW LEMON RESERVOIR (FLOBLECO)
#Retrieved: 1/25/2007 12:06
#-----------------------------------------------------------------------------
Station 	Date/Time	DISCHRG (cfs)
FLOBLECO	2006-01-22 00:00	10.9
FLOBLECO	2006-01-23 00:00	10.7
FLOBLECO	2006-01-24 00:00	10.7

Problems:
Name of the flow column changes from
DISCHRG to DISCHRG1 to DISCHRG2 to DISCHRG3
Only one site at a time can be retrieved
GOES randomly timed data is on a different page than
normal self-timed data?

=cut


if ( defined($readfile) ) {
  my (@data);
  read_from_file( \@data );

  process_data( \@data );

} else {    # attempt to get data from CODWR web page
  foreach my $codwr_no ( keys %$codwr_sites ) {
    my (@data);
    read_from_web( $codwr_no, \@data );

    process_data( \@data, $codwr_no );

    undef(@data);    #needed to reinitialize @data in every loop?
  }    # end of foreach site loop
}

# print error messages for all sites that no or bad data was returned 
my @data_errors;
foreach my $codwr_no ( keys %$codwr_sites ) {
  my $codwr_site = $codwr_sites->{$codwr_no};
  if ( !defined( $codwr_site->{found_data} ) ) {
    push @data_errors,
"No data found for site: $codwr_site->{site_name}, $codwr_no\n";
  } elsif (defined($codwr_site->{error_code})) {
    push @data_errors,
"Bad data seen for site: $codwr_site->{site_name}, $codwr_no: $codwr_site->{error_code}\n";
  }
}

print STDERR sort @data_errors;

exit 0;

#End main program, subroutines below

sub process_args {
  my ( $numdays, $begindate, $enddate, @begindate, @enddate );

  while (@_) {
    my $arg = shift;
    if ( $arg =~ /-h/ ) {    # print help
      &usage;
    } elsif ( $arg =~ /-v/ ) {    # print version
      &version;
    } elsif ( $arg =~ /-t/ ) {    # get test flag
      $insertflag = undef;
    } elsif ( $arg =~ /-f/ ) {    # get file to read from
      $readfile = shift;
      if ( !-e $readfile ) {
        print "file not found: $readfile";
        exit 1;
      }
    } elsif ( $arg =~ /-a/ ) {    # get database login file
      $accountfile = shift;
      if ( !-r $accountfile ) {
        print "file not found or unreadable: $accountfile\n";
        exit 1;
      }
    } elsif ( $arg =~ /-w/ ) {    # get print URL flag
      $printurl = 1;
    } elsif ( $arg =~ /-d/ ) {    # get debug flag
      $debug = 1;
    } elsif ( $arg =~ /-i/ ) {    # get codwr id, split possible id,id,id
      push @site_num_list, split /,/, shift;
    } elsif ( $arg =~ /-u/ ) {    # get hdb user
      $hdbuser = shift;
    } elsif ( $arg =~ /-p/ ) {    # get hdb passwd
      $hdbpass = shift;
    } elsif ( $arg =~ /-n/ ) {    # get number of days
      $numdays = shift();
    } elsif ( $arg =~ /-b/ ) {    # get begin date
      $begindate = shift();
      if ( not @begindate = Decode_Date_EU($begindate) ) {
        print "Error in begin date specified: $begindate\n";
        usage();
      }
    } elsif ( $arg =~ /-e/ ) {    # get end date
      $enddate = shift();
      if ( not @enddate = Decode_Date_EU($enddate) ) {
        print "Error in end date specified: $enddate\n";
        usage();
      }
    } elsif ( $arg =~ /-.*/ ) {    # Unrecognized flag, print help.
      print STDERR "Unrecognized flag: $arg\n";
      usage();
    } else {
      print STDERR "Unrecognized command line argument: $arg\n";
      usage();
    }
  }

  # if user specified codwr gage ids, chop off last comma
  if (@site_num_list) {
    if ( grep ( /[^A-Z]/, @site_num_list ) ) {
      die "ERROR: @site_num_list\ndoes not look like $agen_abbrev id.\n";
    }
  }

  if ( !defined($accountfile) and ( !defined($hdbuser) || !defined($hdbpass) ) )
  {
    print
      "ERROR: Required: accountfile (-a) or HDB username and password(-u -p)\n";
    usage();
  }

  return process_dates( \@enddate, \@begindate, $numdays );
}

sub get_app_ids {

  # Get application wide ids to describe where data came from
  # only need to lookup loading application separately,
  # the data source generic mapping table contain everything but method
  # and computation.

  # defined early in the program
  my $load_app_name = $progname;
  my $sth;

  my $get_load_app_statement = "select loading_application_id from
hdb_loading_application where loading_application_name = '$load_app_name'";

  eval {
    $sth = $hdb->dbh->prepare($get_load_app_statement);
    $sth->execute;
    $sth->bind_col( 1, \$load_app_id );
    unless ( $sth->fetch ) {
      $hdb->hdbdie("Loading application id for $load_app_name not found!\n");
    }
    $sth->fetch;
    $sth->finish();
  };

  if ($@) {    # something screwed up
    print $hdb->dbh->errstr, " $@\n";
    $hdb->hdbdie(
"Errors occurred during selection of loading application id for $load_app_name.\n"
    );
  }

  my $get_other_info_statement = "select e.agen_id,
 e.collection_system_id collect_id, e.data_quality validation,
 e.description url
 from hdb_ext_data_source e
 where e.ext_data_source_name = '$datasource'";

  eval {
    $sth = $hdb->dbh->prepare($get_other_info_statement);
    $sth->execute;
    $sth->bind_col( 1, \$agen_id );
    $sth->bind_col( 2, \$collect_id );
    $sth->bind_col( 3, \$validation );
    $sth->bind_col( 4, \$url );
    my $stuff = $sth->fetch;
    Dumper($stuff);
    unless ($stuff) {
      $hdb->hdbdie(
"Data source definition not found, $agen_id, $collect_id, $validation, $url!\n"
      );
    }
    $sth->finish();
  };

  if ($@) {    # something screwed up
    print $hdb->dbh->errstr, " $@\n";
    $hdb->hdbdie(
       "Errors occurred during selection of data source info for $datasource.\n"
    );
  }

  if ( !defined($validation) ) {

    #needs to be empty string for dynamic sql to recognize as null, not 'null'!
    $validation = '';
  }

}

sub skip_comments {
  my $data = shift;

  # remove leading comments and comments between sites
  while ( substr( $data->[0], 0, 1 ) eq '#' ) {
    shift @$data;
  }

  return ();
}

sub read_header {

  my $data = shift;

  #save header record, for use in finding data column
  my $header = shift @$data;
  my @headers = split /\t/, $header;
  if (!($headers[0] =~ /Station/ )) {
    print "@headers\n";
    $hdb->hdbdie("Data is not $agen_abbrev website tab delimited format!\n");
  }

  #check to see if no data at all, blank line
  if (!defined($data->[0])  
       or ( $data->[0] eq '' )
       or ( substr( $data->[0], 0, 1 ) eq '#' ) )
  {
    shift @$data;
    return 0;    #no data, next station please!
  }

  # now we have data, mark data as found in CODWR ID
  # assume that the format holds, and the id is in second column
  my (@firstrow) = split /\t/, $data->[0];
  my $codwr_no = $firstrow[0];

  # check that codwr_no is sane, i.e. contains only upper case letters:
  if ( $codwr_no =~ /[^[:upper:]]/ ) {
    print STDERR
      "Warning! '$codwr_no' does not appear to be a valid $agen_abbrev site number!\n";
    $hdb->hdbdie("Assuming data retrieved is garbage and giving up!\n");
  }

  #set found data for this codwr no
  $codwr_sites->{$codwr_no}->{found_data} = 1;

  #check that this site is defined completely from database
  if (    !defined( $codwr_sites->{$codwr_no}->{sdi} )
       or !defined( $codwr_sites->{$codwr_no}->{site_id} )
       or !defined( $codwr_sites->{$codwr_no}->{data_code} ) )
  {
    $hdb->hdbdie(
             "site or sdi or data code not defined for codwr id: $codwr_no!\n");
  }

  #find the column in the header that contains the right data code
  my $col = 1;
  until ( $headers[ $col++ ] =~ /$codwr_sites->{$codwr_no}->{data_code}/ ) {
    if ( !defined( $headers[$col] ) ) {
      print STDERR "Not found: '$codwr_sites->{$codwr_no}->{data_code}' in:\n";
      print STDERR "@headers\n";
      print STDERR "Cannot find value column code from header!\n";
      $hdb->hdbdie("Data is not $agen_abbrev website tab delimited format!\n");
    }
  }

  #variable defining which column to read values from
  $codwr_sites->{$codwr_no}->{column} = $col - 1;

  return ($codwr_no);
}

sub process_data {
  my $data           = $_[0];
  my $input_codwr_no = $_[1];

  skip_comments($data);

  my $codwr_no = read_header($data);

  return if $codwr_no eq 0;    # no station found

  if ( defined($input_codwr_no)
       and $input_codwr_no ne $codwr_no )
  {                            #different station found
    $hdb->hdbdie("$input_codwr_no was not returned by web query! Exiting.");
  }

  my $codwr_site = $codwr_sites->{$codwr_no};

  # put data into database, handing site id
  # function returns date of first value and last value updated
  my ( $first_date, $updated_date, $rows_processed );

  if ( defined($insertflag) ) {

    #tell user something, so they know it is working
    print "Working on $agen_abbrev gage number: $codwr_no\n";
    ( $first_date, $updated_date, $rows_processed ) = 
       insert_values( $data, $codwr_site );

#report results of insertion, and report error codes to STDERR
    if ( !defined($first_date) ) {
      print "No data updated for $codwr_site->{site_name}, $codwr_no\n";
    } else {
      print
"Data processed from $first_date to $updated_date for $codwr_sites->{$codwr_no}->{site_name}, $codwr_no\n";
      print "Number of rows from $agen_abbrev processed: $rows_processed\n";
    }
  }
}

sub build_url ($$$) {
  my $site_code  = $_[0];
  my $begin_date = $_[1];
  my $end_date   = $_[2];

#this url is generated from http://www.dwr.state.co.us/help.aspx
# parts of the url:
# program generating the result  : http://www.dwr.state.co.us/SurfaceWater/data/export_tabular.aspx
# specify site                            : ID=$_[0]
# specify discharge(these codes change!)  : MTYPE=$_[1]
# specify instant, hourly, or daily       : INTERVAL=1
# specify start and end dates             : START=YYYY-MM-DD&END=YYYY-MM-DD

  # retrieval from database included data_codes for the various discharge
  # column ids

  die "Site id $_[0] not recognized, no datacode known!\n"
    unless ( defined $codwr_sites->{ $_[0] }->{data_code} );

  my $parameters = "INTERVAL=1&START=$begin_date&END=$end_date";
  $parameters .= "&ID=$_[0]&MTYPE=$codwr_sites->{$_[0]}->{data_code}";

  # url described by database, retrieved in get_app_ids()

  return $url . $parameters;
}

sub build_site_num_list {
  my $codwr_sites = $_[0];
  my ( $codwr_num, @site_num_list );

  #build list of sites to grab data for
  foreach $codwr_num ( keys %$codwr_sites ) {
    push @site_num_list, $codwr_num;
  }

  return @site_num_list;
}

sub build_web_request {
  my $ua = LWP::UserAgent->new;
  $ua->agent(
         "$agen_abbrev Streamflow -> US Bureau of Reclamation HDB dataloader: "
           . $ua->agent );
  $ua->from("$ENV{HDB_XFER_EMAIL}");
  my $request = HTTP::Request->new();
  $request->method('GET');
  return ( $ua, $request );
}

# we appropriate the SCS id field for the CO DWR site ids.
sub get_codwr_sites {
  my $id_limit_clause = '';

  # join arguments with "','" to make quoted string comma delimited list
  if (@_) {
    my $commalist = join( '\',\'', @_ );
    $id_limit_clause = "b.primary_site_code in ( '$commalist' ) and";
  }

  my $get_codwr_no_statement =
    "select b.primary_site_code codwr_id, b.primary_data_code data_code,
 b.hdb_interval_name interval, b.hdb_method_id meth_id,
 b.hdb_computation_id comp_id, b.hdb_site_datatype_id sdi,
 d.site_id, d.site_name
from hdb_site_datatype a, ref_ext_site_data_map b,
hdb_site d, hdb_ext_data_source e
where a.site_datatype_id = b.hdb_site_datatype_id and
b.is_active_y_n = 'Y' and $id_limit_clause 
a.site_id = d.site_id and
b.ext_data_source_id = e.ext_data_source_id and
e.ext_data_source_name = '$datasource'
order by codwr_id";

  $hdb->dbh->{FetchHashKeyName} = 'NAME_lc';

  my $hashref;

  print $get_codwr_no_statement. "\n" if defined($debug);

  eval {
    $hashref = $hdb->dbh->selectall_hashref( $get_codwr_no_statement, 1 );
  };

  if ($@) {    # something screwed up
    print $hdb->dbh->errstr, ": $@\n";
    die "Errors occurred during automatic selection of $agen_abbrev gage ids.\n";
  }

  return $hashref;
}

sub insert_values {
  my @data       = @{ $_[0] };
  my $codwr_site = $_[1];
  my $cur_sdi    = $codwr_site->{sdi};

  #fixed the url encoding to only retrieve one
  # parameter, so no longer need this, but could need it again
  my $column = 1;

  my $numrows    = 0;
  my $first_date = undef;
  my ( $value, $value_date, $updated_date );
  my ( $dummy, $row,        @fields );
  my ( $old_val, $old_source );

  if (    !defined( $codwr_site->{interval} )
       or !defined($agen_id)
       or !defined($load_app_id)
       or !defined($collect_id)
       or !defined( $codwr_site->{meth_id} )
       or !defined( $codwr_site->{comp_id} ) )
  {
    $hdb->hdbdie(
"Unable to write data to database for $codwr_site->{site_name}
Required information missing in insert_values()!\n"
    );
  }

  # SQL statements

  my $modify_data_statement = "
    BEGIN
        modify_r_base(?,'$codwr_site->{interval}',/*sdi, interval, */
                      ?,null,?, /*start_date_time,null end_date_time, value*/
                      $agen_id,null,'$validation',/*agen, overwrite, validation*/
                      $collect_id,$load_app_id,
                      $codwr_site->{meth_id},$codwr_site->{comp_id},
                     'Y',null,                  /*do update?, data flags */
                      HDB_UTILITIES.IS_DATE_IN_DST(?,'CST','MST')); /* timezone for CODWR, they present in localtime */
    END;";

  eval {
    my ($modsth);
    $modsth = $hdb->dbh->prepare($modify_data_statement);

    # insert or update or do nothing for each datapoint;
    foreach $row (@data) {
      $numrows++;
      # throw away first column, which is station id repeated
      ( $dummy, @fields ) = split /\t/, $row;

      $value_date = $fields[0];
      $value      = $fields[$column];

# update or insert, source and database differ (or database value does not exist)
        if ( defined($debug) ) {
	    print "modifying for $cur_sdi, date $value_date, value $value\n";
        }

       # check if value is known
       if ( !defined $value or $value eq '' ) {
         print "data missing: $cur_sdi, date $value_date\n" if defined($debug);
         next;
       } elsif ( $value =~ m/[^0-9\.]/ ) {    # check for other text, complain
         print "data corrupted: $cur_sdi, date $value_date: $value\n";
         $codwr_site->{error_code} = $value;
       } else {       #modify
        $modsth->bind_param( 1, $cur_sdi );
        $modsth->bind_param( 2, $value_date );
        $modsth->bind_param( 3, $value );
        $modsth->bind_param( 4, $value_date );
        $modsth->execute;

        if ( !defined($first_date) ) {    # mark that data has changed
          $first_date = $value_date;
        }
        $updated_date = $value_date;
      } 
    }
    $modsth->finish;

  };    # semicolon here because of use of eval

  
  if ($@) {    # something screwed up in insert/update
    print "$hdb->dbh->errstr, $@\n";
    $hdb->dbh->rollback;
    $hdb->hdbdie(
"Errors occurred during insert/update/deletion of data. Rolled back any database manipulation.\n");
  } elsif ($first_date) {    # commit the data
    if ( defined($debug) ) {
      $hdb->dbh->rollback or $hdb->hdbdie($hdb->dbh->errstr);
    } else {
      $hdb->dbh->commit or $hdb->hdbdie($hdb->dbh->errstr);
    }
  }
  return ( $first_date, $updated_date, $numrows );
}

sub usage {
  print STDERR <<"ENDHELP";
$progname [ -h | -v ] | [ options ]
Retrieves $agen_abbrev flow data and inserts into HDB.
Example: $progname -u app_user -p <hdbpassword> -i AZOTUNNM

  -h               : This help
  -v               : Version
  -u <hdbusername> : HDB application user name (REQUIRED)
  -p <hdbpassword> : HDB password (REQUIRED)
  -i <codwr_id>    : $agen_abbrev gage id to look for. Must be in HDB. More than one
                     may be specified with -i id1,id2 or -i id1 -i id2
                     If none specified, will load all gages set up in HDB.
  -t               : Test retrieval, but do not insert data into DB
  -f <filename>    : Filename instead of live web
  -w               : Web address (URL developed to get $agen_abbrev data)
  -d               : Debugging output
  -n               : number of days to load
  -b <DD-MON-YYYY> : Begin date for data retrieval
  -e <DD-MON-YYYY> : End date for data retrieval
ENDHELP

  exit(1);
}

sub version {
  my $verstring = '$Revision$';
  $verstring =~ s/^\$Revision: //;
  $verstring =~ s/\$//;

  print "$progname version $verstring\n";
  exit 1;
}

sub read_from_file {
  my $data = shift;

  open INFILE, "$readfile" or $hdb->hdbdie("Cannot open $readfile: $!\n");
  @$data = <INFILE>;
  print @$data         if defined($debug);
  print Dumper(@$data) if defined($debug);
  chomp @$data;
  return ();
}

sub read_from_web {
  my $codwr_no = shift;
  my $data     = shift;

  my $url = build_url( $codwr_no, $begindatestr, $enddatestr );

  if ( defined($printurl) or defined($debug) ) {
    print "$url\n";
  }

  my ( $ua, $request ) = build_web_request();
  $request->uri($url);

  # this next function actually gets the data
  my $response = $ua->simple_request($request);

  # check result
  if ( $response->is_success ) {
    my $status = $response->status_line;
    print "Data read from $agen_abbrev: $status\n";

    if ( !$response->content ) {
      $hdb->hdbdie("no data returned from $agen_abbrev, exiting.\n");
    }
    print $response->content if defined($debug);

    @$data = split /$INPUT_RECORD_SEPARATOR/, $response->content;
  } else {    # ($response->is_error)
    printf STDERR $response->error_as_HTML;
    $hdb->hdbdie("Error reading from $agen_abbrev web page, cannot continue\n");
  }
  return;
}

sub process_dates {
  my $enddate   = shift;
  my $begindate = shift;
  my $numdays   = shift;

  #Truth table and results for numdays, begindate and enddate
  # errors are all defined, only enddate defined.
  # everything else gets fixed up.
  # Num Days | Beg Date | End Date | Result
  # nothing defined                   error
  #  def                              end date = now, beg date = enddate-numdays
  #  def        def                   end date = beg date + numdays
  #             def                   end date = now, check max num days?
  #             def         def       ok
  #                         def       error, what is beg date?
  #  def                    def       beg date = end date - numdays
  #  def        def         def       error, not checking to see if consistent
  #
  #  We end up with a begin date and an end date
  if ( defined($numdays) ) {
    if ( $numdays =~ m/\D/ ) {
      print "Number of days (argument to -n) must be numeric.\n";
      usage();
    }

    if ( @$begindate and @$enddate ) {
      print "Error, overspecified dates, all of -n, -b, -e specified!\n";
      exit 1;
    } elsif (@$begindate) {
      @$enddate = Add_Delta_Days( @$begindate, $numdays );
    } elsif (@$enddate) {
      @$begindate = Add_Delta_Days( @$enddate, -$numdays );
    } else {
      @$enddate = Today();
      @$begindate = Add_Delta_Days( @$enddate, -$numdays );
    }
  } else {
    if ( @$begindate and @$enddate ) {

      #do nothing, we're good
    } elsif (@$begindate) {
      @$enddate = Today();
    } elsif (@$enddate) {
      print "Error, cannot specify only end date!\n";
      exit 1;
    } else {
      print "Error, dates are unspecified!\n";
      usage();
    }
  }

  if ( !@$begindate or !@$enddate ) {
    print "Error, dates not specifed or error parsing dates!\n";
    exit(1);
  }

  my $begindatestr = sprintf( "%4d-%02d-%02d", @$begindate );
  my $enddatestr   = sprintf( "%4d-%02d-%02d", @$enddate );

  return ( $begindatestr, $enddatestr );
}
