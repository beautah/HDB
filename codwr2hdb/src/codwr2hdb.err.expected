Bad data seen for site: DOVE CREEK CANAL BELOW GREAT CUT DIKE, DOVCANCO: E
